src/main/cpp/src/bloom_filter.cu

```diff
@@ -0,0 +1,153 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "murmur_hash.cuh"
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/detail/utilities/cuda.cuh>
+#include <cudf/null_mask.hpp>
+#include <cudf/utilities/bit.hpp>
+#include <cudf/utilities/span.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+namespace spark_rapids_jni {
+
+namespace {
+
+using bloom_hash_type = spark_rapids_jni::murmur_hash_value_type;
+
+__device__ inline std::pair<cudf::size_type, cudf::bitmask_type> gpu_get_hash_mask(
+  bloom_hash_type h, cudf::size_type bloom_filter_bits)
+{
+  // https://github.com/apache/spark/blob/7bfbeb62cb1dc58d81243d22888faa688bad8064/common/sketch/src/main/java/org/apache/spark/util/sketch/BloomFilterImpl.java#L94
+  auto const index      = (h < 0 ? ~h : h) % static_cast<bloom_hash_type>(bloom_filter_bits);
+  auto const word_index = cudf::word_index(index);
+  auto const bit_index  = cudf::intra_word_index(index);
```

Comment:
This almost works, but the data format we receive from Spark is actually big endian, because we receive the serialized form of the Bloom filter (see BloomFilterImpl.writeTo which uses DataOutputStream which in turn encodes big-endian).  We can operate on the big-endian longs directly like this:
```suggestion
  auto const word_index = cudf::word_index(index) ^ 0x1;
  auto const bit_index  = cudf::intra_word_index(index) ^ 0x18;
```

Reply:
Done